### PR TITLE
NIFIREG-321 Integrate revision concept into UI, REST API, and service…

### PR DESCRIPTION
**Pull request by bbende**
… layers.

This PR has a significant refactoring to some of the REST resources and service layer. The idea is for all REST resources to call the ServiceFacade which is the entry point for a database transaction. The ServiceFacade then applies authorization, validation, and revision checking. 

This also allows us to remove the locking mechanisms in the services, since the locking will now be based on optimistic-locking in the database.

A follow on PR will add database implementations of UserGroupProvider and AccessPolicyProvider, which when combined with the DatabaseFlowPersistenceProvider and the metadata database, will provide a way for all data to be managed in the database as part of a single transaction.

### CODE REVIEW

**Comment by bbende** [comment]
I also wanted to add that in terms of testing this, the objects that support revisions are bucket, flow, user, user group, and policy. 

You can have two different browsers open and perform tests where you open the same object for editing in both browsers, and submit an edit in one, and then the edit should be rejected in the other since the state is now stale.

The user/group/policy parts are the most complex due to the various ways which the UI allows for editing these entities, so there are several variations to test.

**Comment by kevdoran** [comment]
@bbende Thanks for this contribution. This will be an important addition for HA deployment scenarios. I have taken a quick look at the code changes. I have not built or run this branch yet, but I wanted to provide you an update and initial feedback.

I have reservations about adding the clientId field to the `RevisionInfo` object for NiFi Registry DTOs. I am aware that this is is being brought over from NiFi and that this design is consistent with NiFi. I am not aware of what value it adds to NiFi Registry (will our frontend or the nifi client use it?), and I think it comes with potential downsides:

1. If the plan for `RevisionInfo` is (at least partially) to allow for optimistic locking and prevent lost/overwritten updates in a concurrent access situation, then misuse (intentional or accidental) of the clientId field allows for overriding another client's update.
2. It adds complexity and confusion to the REST API documentation and correct usage for newcomers. A recent example of this from the nifi side: https://apachenifi.slack.com/archives/C0L9VCD47/p1570538029297700

If the plan is that the new library modules introduced by this PR could eventually move to `nifi-standard-libraries` to be shared across `nifi` and `nifi-registry`, then I am in support of having an _optional_ clientId field that the library understands, so that it can be used in the nifi case. However, I don't think NiFi Registry should be taking advantage of it in its DTOs.

**Comment by bbende** [comment]
@kevdoran that is fair, I think most likely we could just remove clientId from the `RevisionInfo` DTO and not much else should change besides one or two places in the code that create `RevisionInfo` and convert to the internal Revision object.

**Comment by bbende** [comment]
@kevdoran I pushed another commit that removes the use of clientId in the REST API, but I am still on the fence about removing it. Here are some reasons...

1) The revision/optimistic-locking is not really meant to be a security concept, it is just to prevent modifications from stepping on each other. The user still has to be authenticated and authorized to perform the given action on the given resource, which is really whats protecting the system. 

2) Even without clientId, anyone can still misuse the version field. If someone knows the last version was 10, they can send any update they want with version as 10. It doesn't prove that the update they sent is actually based off version 10, they could have crafted any entity they wanted and then set version to 10.

3) Consistency could be nice. While I do agree that the API is complicated, it is also nice that it works the same across the two systems. 

4) The `RevisionInfo` class is part of `nifi-registry-revision-entity-model` and was meant to be a reusable DTO for any consumers that want to add this concept into their web layer, and the `RevisableEntityService` operated on `RevisableEntity`, so now it means that the entity service part of the library doesn't really support clientId anymore.

This might be fine because the idea of the library was the consumers could pick and choose which parts to use, and I don't think NiFi would use that part anyway since it has it's own `RevisionDTO`, but I was hoping that I wouldn't have to change anything inside the revision modules for this change.  I suppose I could add another DTO to shield registry objects from RevisionInfo and RevisableEntity, but seems like overkill.

**Comment by kevdoran** [comment]
Hey @bbende, those are all good points. I’m good with leaving it. Thanks for taking the time to explain your thinking and the intent to me. 

I’ll finish my review ASAP. Thanks!

File: nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/service/StandardServiceFacade.java [hunk 1]
```diff
@@ -0,0 +1,1099 @@
...
+    public static final String BUCKET_ENTITY_TYPE = "Bucket";
+    private final RegistryService registryService;
+    private final ExtensionService extensionService;
+    private final AuthorizationService authorizationService;
+    private final AuthorizableLookup authorizableLookup;
+    private final RevisableEntityService entityService;
+    private final PermissionsService permissionsService;
+    private final LinkService linkService;
+
+    public StandardServiceFacade(final RegistryService registryService,
```
[the remark this answers is not in the document]
Comment (bbende) [reply nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/service/StandardServiceFacade.java]: Good call, I'll update this to be consistent

File: nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/mapper/InvalidRevisionExceptionMapper.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public Response toResponse(InvalidRevisionException exception) {
+        // log the error
+        logger.info(String.format("%s. Returning %s response.", exception, Response.Status.BAD_REQUEST));
+
+        if (logger.isDebugEnabled()) {
+            logger.debug(StringUtils.EMPTY, exception);
+        }
+
+        return Response.status(Response.Status.BAD_REQUEST).entity(exception.getMessage()).type("text/plain").build();
```
[the remark this answers is not in the document]
Comment (bbende) [reply nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/mapper/InvalidRevisionExceptionMapper.java]: 1) I don't mind changing it to 409 if you think that is more appropriate. I think I just brought over the same `InvalidRevisionExceptionMapper` that NiFi has which uses 400. 

2) I agree that the error message is not really helpful for the end user. I think something like what you said in the general comments probably makes sense for a more friendly message. 

3) `ExpiredRevisionClaimException` is a sub-class of `InvalidRevisionException` so it is getting handled by that exception mapper. I will be honest and say that I am not sure why `ExpiredRevisionClaimException ` even exists. It was only used during a delete in the` NaiveRevisionManager`, and its the same check that is done in update where we use `InvalidRevisionException`, and as far as I can tell NiFi doesn't have any specific logic to handle it differently, in fact there are two places where it catches it and converts it to `InvalidRevisionExcpetion `. So long story short, I left it in place for consistency, but I think the error handling is primarily based on `InvalidRevisionException`.

**Comment by bbende** [comment]
@kevdoran thanks for the review! I'll work on some updates, but may take me a couple of days to get back to it.

For the general comments...

1) I think the idea of checking the revision on delete is so the client knows that they might be deleting something that isn't in the state they think it is in. I can't come up with a good real world example, but something like... user1 goes to delete a bucket named "Not used", but user2 has already changed the name to "Do not delete", so user1 gets a revision exception, refreshes, and then realizes they shouldn't be deleting the bucket.  

    As far as getting the InvalidRevisionException first, that is mostly because of how the revision checking is implemented by wrapping whatever the operation is, so that check happens first before the actual delete function. This also may tie into whether we want to use 400 or 409, since 409 seems more appropriate for cases where the entity exists but the revision isn't correct, but since we have other cases where there literally is no entity with the given revision, maybe that is more of a 400.

2) I'll look into updating the messages coming back.

3) I think I went back and forth on how to best handle this, and you are definitely right that there are some inconsistencies (i.e. bucket vs add user to group). At first I thought it seemed wrong to lose changes the user made and replace the entity with the latest from the server, but since there is really no way for the user to resolve the issues without doing a refresh on their own, then I think we should probably try to do that when we can. For this one I'd say maybe we can make a follow on JIRA to go through and try and make it consistent.

**Comment by kevdoran** [comment]
Thanks @bbende. Take your time on making any changes, I know this PR has been up a while, so no rush context-switching to get back to it. 

Regarding the ongoing points we are discussing:

1. The example you gave, someone renaming something to "do not delete" or "WIP" or something like that is a good counter example. I don't really have a problem with checking revision on delete as it doesn't really do any harm to check, and it could catch something worth preventing in cases we cannot think of, so let's leave that as is.

    Regarding the 400/409 issue, ideally/conceptually, my preference would be to following:

        Entity does not exist: 404 is returned for all operations (GET, PUT, DELETE)
        Entity exists, revision is stale: 409 
        Any other reason / malformed request / failed revision check / etc: 400

    That said, I think this needs to be balanced with feasibly. I know in some cases in the JDBC impl that the exception is being thrown based on the results of executing a SQL statement, and if we think that result could be due to multiple reasons that are difficult to determine, then I would be ok using a 400 as a catch all for "failed do to one of many potential reasons".

2. Sounds good.

3. Good with this as a follow on Jira. Thanks!

File: nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/mapper/InvalidRevisionExceptionMapper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-core/nifi-registry-web-api/src/main/java/org/apache/nifi/registry/web/mapper/InvalidRevisionExceptionMapper.java]: Thanks for the clarification about `ExpiredRevisionClaimException`. I did not realize it extended `InvalidRevisionException`

**Comment by bbende** [comment]
@kevdoran I just pushed two new commits...

The first one makes the user message more friendly for `InvalidRevisionException`, the message now is: 
`
The %s you attempted to %s with id '%s' is out of date with the server. You may need to refresh your client and try again.`

I included the id only because there are cases where a multi-delete can be performed and its hard to know which entity has the issue.

The second commit has some frontend updates to try and make the error handling consistent when encountering an invalid revision. Everything should now be trying to refresh the entity automatically. There are a few weird cases where the entity was deleted and you'll get the revision error and then also another not found error. We could possibly improve some of this in a future effort, but I think it is acceptable for now given these are mostly edge cases.

I decided to leave the response code as a 400 because the frontend already had some code where it expected a 409 to be specifically an issue with a unique name for a bucket/user/group, plus it isn't really clear in the JDBC case as to the reason why no rows were updated (version not current, no revision for the given entity id).

Let me know if there is anything else that needs to be updated. Hopefully we can get this in and then can rebase the PR #232.

**Comment by bbende** [comment]
@kevdoran pushed one more commit that addresses one of your comments about getting a 400 when the entity didn't exist and should have been a 404. 

The change adds a check for existence in the service facade during update or delete of a `RevisableEntity`, before entering the revision checking. This way if the entity doesn't exist it will throw a `ResourceNotFoundException` (404), instead of `InvalidRevisionException`.

There is a possible race condition where the existence check could pass, and before entering the revision checking, the entity then gets deleted. This would be extremely rare and would produce an `InvalidRevisionException`, at which point the client refreshes and the entity would be gone.

**Pull request closed without merging**